controller/sync.go

```diff
@@ -277,23 +279,49 @@ func (sc *syncContext) sync() {
 	// if it is the last phase/wave and the only remaining tasks are non-hooks, the we are successful
 	// EVEN if those objects subsequently degraded
 	// This handles the common case where neither hooks or waves are used and a sync equates to simply an (asynchronous) kubectl apply of manifests, which succeeds immediately.
-	complete := tasks.Find(func(t *syncTask) bool { return t.phase != phase || wave != t.wave() || t.isHook() }) == nil
+	complete := !tasks.Any(func(t *syncTask) bool { return t.phase != phase || wave != t.wave() || t.isHook() })
 
-	sc.log.WithFields(log.Fields{"phase": phase, "wave": wave, "tasks": tasks}).Debug("filtering tasks in correct phase and wave")
+	sc.log.WithFields(log.Fields{"phase": phase, "wave": wave, "tasks": tasks, "syncFailTasks": syncFailTasks}).Debug("filtering tasks in correct phase and wave")
 	tasks = tasks.Filter(func(t *syncTask) bool { return t.phase == phase && t.wave() == wave })
 
 	sc.setOperationPhase(v1alpha1.OperationRunning, "one or more tasks are running")
 
 	sc.log.WithFields(log.Fields{"tasks": tasks}).Debug("wet-run")
 	if !sc.runTasks(tasks, false) {
-		sc.setOperationPhase(v1alpha1.OperationFailed, "one or more objects failed to apply")
+		sc.runFailedTasksIfAny(syncFailTasks, "one or more objects failed to apply")
 	} else {
 		if complete {
 			sc.setOperationPhase(v1alpha1.OperationSucceeded, "successfully synced (all tasks run)")
 		}
 	}
 }
 
+func (sc *syncContext) runFailedTasksIfAny(syncFailTasks syncTasks, failMessage string) {
+	if len(syncFailTasks) > 0 {
+		// If tasks are already completed, don't run them again
+		if syncFailTasks.All(func(task *syncTask) bool { return task.completed() }) {
+			if syncFailTasks.All(func(task *syncTask) bool { return task.successful() }) {
+				sc.setOperationPhase(v1alpha1.OperationFailed, failMessage)
+				syncFailTasks.ApplyResourceResult(sc, v1alpha1.ResultCodeSynced, v1alpha1.OperationSucceeded, "applied successfully")
+			} else {
+				sc.setOperationPhase(v1alpha1.OperationFailed, failMessage)
+				successfulTasks, failedTasdks := syncFailTasks.Split(func(task *syncTask) bool {
+					return task.successful()
+				})
+				successfulTasks.ApplyResourceResult(sc, v1alpha1.ResultCodeSynced, v1alpha1.OperationSucceeded, "applied successfully")
+				failedTasdks.ApplyResourceResult(sc, v1alpha1.ResultCodeSynced, v1alpha1.OperationFailed, "failed to apply")
+			}
+			return
+		}
+		sc.log.WithFields(log.Fields{"syncFailTasks": syncFailTasks}).Debug("running sync fail tasks")
+		if !sc.runTasks(syncFailTasks, false) {
+			sc.setOperationPhase(v1alpha1.OperationFailed, failMessage)
+		}
+	} else {
+		sc.setOperationPhase(v1alpha1.OperationFailed, failMessage)
```

Comment:
I think you should do this at the start of the function.

Reply:
I don't think that failing to run any of the tasks (including panic) should affect this status.

Reply:
This actually can't be at the beginning of the function. The operation is technically still `OperationRunning` until all of the `SyncFail` Hooks are done. Since this function triggers the `SyncFail` hooks to begin with, setting the operation to `OperationFailed` will prematurely end it and the fail hooks will not be run